### PR TITLE
Add taskparameter type

### PR DESCRIPTION
**Pull request by Tarrasch**
Not done yet, I'll try to move over Range to use this before merging, just to see that it works in practice too. :)

### CODE REVIEW

**Comment by sisidra** [comment]
I suggest to remove Luigi custom namespace in favour of python package.module first and then task modules are not required to be imported first (can do it at runtime).
Otherwise this to work requires import everything. I don't see it as good practice.

**Comment by Tarrasch** [comment]
Seems sensible, at least to make it a default. We discussed this earlier (forgot ticket number).

Anyway, this patch is not directly related to that. :)

**Comment by Tarrasch** [comment]
Ok, please review folks! :)

@sisidra @ulzha @erikbern

File: luigi/parameter.py [hunk 1]
```diff
@@ -552,3 +553,26 @@ def parse(self, input):
...
+    When used programatically, the parameter should be specified
+    directly with the :py:class:`luigi.task.Task` (sub) class. Like
+    ``MyMetaTask(my_task_param=my_tasks.MyTask)``. On the command line,
+    you specify the :py:attr:`luigi.task.Task.task_family`. Like
+
+    .. code:: console
+
+            $ luigi --module my_tasks MyMetaTask --my_task_param my_namespace.MyTask
+
+    When instantiated, it'll always instantiate to a task class.
```
Comment (sisidra) [inline luigi/parameter.py]: Add note that `MyTask`'s module *must* be imported in `my_tasks` module

**Comment by sisidra** [comment]
I hold my previous comment about referencing task by python package.module.class name not Luigi's one.

File: luigi/parameter.py (hunk 1 above)
Comment (Tarrasch) [inline luigi/parameter.py]: Good point! :)

**Comment by Tarrasch** [comment]
> I hold my previous comment about referencing task by python package.module.class name not Luigi's one.

I agree that it makes sense in many settings. But I'm not sure we should remove the support for it, hence forcing that new convention. I'm afraid it have implications we haven't fully thought about. We could however:

 * Make it as a configurable default to set the `task_namespace` to be the same as package+module. Big organizations can then turn this on by default in config then.
 * Extend `TaskParameter` to silently "try" do the module loading. If we also change the current command line parsing (which I want to do anyway), making the `--task` param a `TaskParameter`, we would allow users to shorten cron lines from `--module my_module --task my_module.MyTask`  to only `--task my_module.MyTask`.

What do you think about that? I think we could keep the `task_namespace` concept in luigi though for those who want it.

**Comment by Tarrasch** [comment]
@sisidra Does this look better now? :)

File: luigi/parameter.py [hunk 2]
```diff
@@ -552,3 +553,29 @@ def parse(self, input):
...
+    When used programatically, the parameter should be specified
+    directly with the :py:class:`luigi.task.Task` (sub) class. Like
+    ``MyMetaTask(my_task_param=my_tasks.MyTask)``. On the command line,
+    you specify the :py:attr:`luigi.task.Task.task_family`. Like
+
+    .. code:: console
+
+            $ luigi --module my_tasks MyMetaTask --my_task_param my_namespace.MyTask
+
+    Where ``my_namespace.MyTask`` is defined in the ``my_tasks`` python module.
```
Comment (sisidra) [inline luigi/parameter.py]: It should be in registry, not only defining but also importing works :P

**Pull request merged**